Core.ApplicationServices/Users/Write/UserWriteService.cs

```diff
@@ -216,29 +222,44 @@ private Result<User, OperationError> ChangeLocalAdminStatus<T>(Guid organization
                         transaction.Commit();
                         return user.Value;
                     },
-                    failure => { 
+                    failure =>
+                    {
                         transaction.Rollback();
-                        return new OperationError(failure); 
+                        return new OperationError(failure);
                     });
         }
 
-        private Result<User, OperationError> SetUsersGlobalAdminStatus(Guid userUuid, bool status)
+        private Result<User, OperationError> UpdateUser(Guid userUuid, Func<User, Result<User, OperationError>> mutation)
         {
             using var transaction = _transactionManager.Begin();
-            return _userService.GetUserByUuid(userUuid)
-                .Bind(user => UpdateGlobalAdminStatus(user, status))
-                .Match<Result<User, OperationError>>(
-                    user =>
-                    {
-                        transaction.Commit();
-                        _userService.UpdateUser(user, null, null, true);
-                        return user;
-                    },
-                    error =>
-                    {
-                        transaction.Rollback();
-                        return error;
-                    });
+            var updateResult = _userService.GetUserByUuid(userUuid)
+                .Bind(mutation);
+            if (updateResult.Failed)
+            {
+                transaction.Rollback();
+                return updateResult.Error;
+            }
+
+            var updatedUser = updateResult.Value;
+            _userService.UpdateUser(updatedUser, null, null, true);
+            transaction.Commit();
+            return updatedUser;
+        }
+
+        private Result<User, OperationError> SetUsersGlobalAdminStatus(Guid userUuid, bool status)
+        {
+            return UpdateUser(userUuid, (user) => UpdateGlobalAdminStatus(user, status));
+        }
+
+        private Result<User, OperationError> UpdateSystemIntegrator(User user, bool systemIntegratorStatus)
+        {
+            if (!_organizationalUserContext.IsGlobalAdmin())
+            {
+                return new OperationError("You do not have the permission to change this role", OperationFailure.Forbidden);
+            }
+
+            user.IsSystemIntegrator = systemIntegratorStatus;
```

Comment:
I think this assignment should be moved to the Domain model (UpdateSystemIntegrator)

The same thing would be nice for IsGlobalAdmin.

It wasn't done before we took over, and that's why a big chunk of the app isn't implemented like this, but we try to stick to the rule that update logic, even simple, belongs to the Domain. (See SystemUsage for good examples)

Reply:
And then if you move this to the domain, you can extract the IsGlobalAdmin check to a common method (it's reused here and in the UpdateGlobalAdminStatus), verify the call before calling the UpdateUser (or pass it into the UpdateUser, like the mutation), and that will allow you to remove the UpdateSystemIntegrator method - a bit code less. 

Consider doing the same with the UpdateGlobalAdminStatus